service/reportsStorage/storage.go

```diff
@@ -0,0 +1,361 @@
+	storageModels "github.com/iryonetwork/wwm/gen/storage/models"
+	"github.com/iryonetwork/wwm/storage/s3"
+	"github.com/iryonetwork/wwm/storage/s3/object"
+)
+
+// Service describes storage service's public API.
+type Service interface {
+	// Checksum calculates the checksum of a given reader using sha256.Sum.
+	Checksum(r io.Reader) (string, error)
+
+	// ReportList returns a list of latest versions of files. Older versions and
+	// files marked as deleted are removed from the list.
+	ReportList(ctx context.Context, reportType string) ([]*models.ReportFileDescriptor, error)
+
+	// ReportGet returns the latest version of the file by returning the reader
+	// and file details.
+	ReportGet(ctx context.Context, reportType, fileName string) (io.ReadCloser, *models.ReportFileDescriptor, error)
+
+	// ReportGetVersion returns a specific version of a file.
+	ReportGetVersion(ctx context.Context, reportType, fileName, version string) (io.ReadCloser, *models.ReportFileDescriptor, error)
+
+	// ReportListVersions returns a list of all modifications to a file.
+	ReportListVersions(ctx context.Context, reportType, fileName string, createdAtSince, createdAtUntil *strfmt.DateTime) ([]*models.ReportFileDescriptor, error)
+
+	// ReportNew creates a new file.
+	ReportNew(ctx context.Context, reportType string, r io.Reader, contentType string, dataSince *strfmt.DateTime, dataUntil strfmt.DateTime) (*models.ReportFileDescriptor, error)
+
+	// ReportUpdate creates a new version of a file.
+	ReportUpdate(ctx context.Context, reportType, fileName string, r io.Reader, contentType string) (*models.ReportFileDescriptor, error)
+
+	// ReportDelete marks file as deleted.
+	ReportDelete(ctx context.Context, reportType, fileName string) error
+}
+
+// Bucket or item was already deleted
+var ErrDeleted = s3.ErrDeleted
+
+// Item does not exists
+var ErrNotFound = s3.ErrNotFound
+
+// Item already exists
+var ErrAlreadyExists = s3.ErrAlreadyExists
+
+type service struct {
+	s3     s3.Storage
+	logger zerolog.Logger
+}
+
+func (s *service) Checksum(r io.Reader) (string, error) {
+	h := sha256.New()
+	if _, err := io.Copy(h, r); err != nil {
+		return "", err
+	}
+
+	return base64.URLEncoding.EncodeToString(h.Sum(nil)), nil
+}
+
+func (s *service) ReportList(ctx context.Context, reportType string) ([]*models.ReportFileDescriptor, error) {
+	// init list to return
+	list := []*models.ReportFileDescriptor{}
+
+	// check if bucket exists
+	exists, err := s.s3.BucketExists(ctx, reportType)
+	if err != nil {
+		s.logger.Info().Err(err).Str("bucket", reportType).Msg("Failed to check if bucket exists")
+		return nil, err
+	}
+	if !exists {
+		return list, nil
+	}
+
+	// collect the list
+	l, err := s.s3.List(ctx, reportType, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// extract only latest versions; latest version is already sorted
+	// create ReportFileDescriptor objects
+	m := map[string]bool{}
+	for _, f := range l {
+		if _, ok := m[f.Name]; !ok {
```

Comment:
invalid operation: f (variable of type *invalid type) has no field or method Name